File: LeetCode/word_search.rb
grid = [["A", "B", "C", "E"],
        ["S", "F", "C", "S"],
        ["A", "D", "E", "E"]]

def search(s, grid)
    0.upto(grid.size - 1).each { |i|
        0.upto(grid[0].size - 1). each { |j|
            return true if search_rec(grid, s, i, j)
        }
    }
    
    return false
end

def search_rec(grid, s, i, j)
    return false if !i.between?(0, grid.size - 1) || !j.between?(0, grid[0].size - 1)
    return false if s.first != grid[i][j]
    return true if s.size == 1
    aux = grid[i][j]
    grid[i][j] = nil
    result = search_rec(grid, s[1..-1], i + 1, j) ||
             search_rec(grid, s[1..-1], i, j + 1) || 
             search_rec(grid, s[1..-1], i - 1, j) || 
             search_rec(grid, s[1..-1], i, j - 1)
    grid[i][j] = aux
    return result
end

["ABCCED", "SEE", "ABCB"].each { |x|
    puts "#{x} #{search(x.chars, grid)}"
}

File: LeetCode/word_search.cs
using System;

class Program {
    
    static bool Search(char[][] grid, int line, int column, String word, int position) {
        if (!(0 <= line && line < grid.Length && 0 <= column && column < grid[0].Length)) {
            return false;
        }

        char current = grid[line][column];

        if (current != word[position]) {
            return false;
        }
        
        if (position == word.Length - 1) {
            return true;
        }

        grid[line][column] = '*';
        bool result = (Search(grid, line + 1, column, word, position + 1) ||
                       Search(grid, line - 1, column, word, position + 1) ||
                       Search(grid, line, column + 1, word, position + 1) ||
                       Search(grid, line, column - 1, word, position + 1));

        grid[line][column] = current;
        return result;
    }

    static void Main() {
        char[][] grid = new char[][] { 
            new char[4] {'a', 'b', 'c', 'e' },
            new char[4] {'s', 'f', 'c', 's' },
            new char[4] {'a', 'd', 'e', 'e' }
        };

        string[] words = new string[] { "abcced", "see", "abcb" };
        foreach (string word in words) { 
            bool found = false;
            for (int i = 0; i < grid.Length && !found; i++) {
                for (int j = 0; j < grid[0].Length && !found; j++) {
                    if (Search(grid, i, j, word, 0)) {
                        Console.WriteLine("Found {0}", word);
                        found = true;
                    }
                }
            }

            if (!found) {
                Console.WriteLine("Not found {0}", word);
            }
        }
    }
}
